h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -906,17 +906,43 @@ protected boolean ignoreUuidColumns() {
    */
   protected void ignoreBadColumns(int npredictors, boolean expensive){
     // Drop all-constant and all-bad columns.
-    if(_parms._ignore_const_cols)
+    if(_parms._ignore_const_cols && _parms._checkpoint == null) {
       new FilterCols(npredictors) {
-        @Override protected boolean filter(Vec v) {
+        @Override
+        protected boolean filter(Vec v) {
           boolean isBad = v.isBad();
           boolean skipConst = ignoreConstColumns() && v.isConst();
           boolean skipString = ignoreStringColumns() && v.isString();
           boolean skipUuid = ignoreUuidColumns() && v.isUUID();
           boolean skip = isBad || skipConst || skipString || skipUuid;
           return skip;
         }
-      }.doIt(_train,"Dropping bad and constant columns: ",expensive);
+      }.doIt(_train, "Dropping bad and constant columns: ", expensive);
```

Comment:
I think we should drop bad columns and columns the algo doesn't support (strings).

If the user changes the type of the column (eg. from Number to UUID) then this will fail in the model training. I think this needs a little different approach - but you are on the right path.

Reply:
This is old code you are probably referring to. 

The code below only deletes features not present in the checkpointed model. That's it. I can see the limits of this approach. The goals are:

- Implement approach checking column types of shared columns being equal 
- Use list of columns in the checkpointed model as a blacklist - if the same column is in the original dataset, it can not be blacklisted when resuming from that checkpoint (the root cause for this error we're trying to solve here).
- Do not delete new columns, as it might make sense for some algos to leverage them  ? For example DRF (trees may be added with new splits on features unseen in previously created trees ?).

Reply:
This is all a matter of how we define our operations. Let `f(x)` be a model-creating function. Also, let `A` be a training set. Let's consider a model `M = f(A)`.

Now, split the training set `A` into two subsets `B,C`, so that `B + C = A`. Creating a checkpointed model by `Mc = f(B)` and then finishing the model from where we left off by `Mf = f(C, Mc)`, the resulting model `Mf` should be equal to the original `M`. (`M = Mf`).

If these models should be equal, than we may not drop bad columns in all cases, as a column from a subset of the original data might look like a bad column, bad when the whole original dataset is fed, the column no longer looks "bad", e.g. being full of 0s. This is exactly the case with this PR.